### DISCUSSION TITLE
How to delete all the links created after migrations except Package links and their links?

### DISCUSSION THREAD

**Discussion by FreePhoenix888** [deep-foundation/Discussions#15] [Q&A]
How to delete all the links created after migrations except Package links and their links?

**Comment by konard** [comment]
Well it should be quite easy because now you have MigrationsEnd link instance that marks the end of transactions.

**Reply by FreePhoenix888** [reply]
Your answer still do not answer my question. For example this code
```ts
await deep.delete({
   up: {
      tree_id: {
         _eq: await deep.id("@deep-foundation/core", "containTree")
      },
      parent: {
         id: {
            _gt: migrationsEndLink.id,
         },
         type_id: await deep.id("@deep-foundation/core", "Contain"),
         from_id: containerLinkId,
      }
   },
})
```
Deletes ALL the links created after migrations end. Where should I add _not to not delete Package and all the links contained in the package?

**Reply by konard** [reply]
No, it does not. It only deletes something connected to container inside containTree.

I recommend to start from something like this
```js
await deep.delete({
  id: {
     _gt: migrationsEndLink.id,
  },
},
```

Then add `_not` condition, and so on.

```js
await deep.delete({
  id: {
     _gt: migrationsEndLink.id,
  },
  _not: {
     type_id: await deep.id("@deep-foundation/core", "Package"),
  }
},
```

Or something like this to keep complete package structure intract

```js
await deep.delete({
  id: {
     _gt: migrationsEndLink.id,
  },
  _not: {
     up: {
        tree_id: {
           _eq: await deep.id("@deep-foundation/core", "containTree")
        },
        parent: {
            type_id: await deep.id("@deep-foundation/core", "Package"),
        }
     },
  }
},
```

Be sure to also keep PackageNamespace, PackageActive, PackageVersion links and so on.

**Reply by FreePhoenix888** [reply]
> No, it does not. It only deletes something connected to container inside containTree.

I have just tried it and after executing my code I find out that my package is installed.
Try to install a package then execute my code then try to find that package by using deep.id

**Reply by FreePhoenix888** [reply]
> ```js
> await deep.delete({
>   id: {
>      _gt: migrationsEndLink.id,
>   },
>   _not: {
>      type_id: await deep.id("@deep-foundation/core", "Package"),
>   }
> },
> ```
This code will not delete only package links, but the links contained in the package will be deleted

**Reply by FreePhoenix888** [reply]
> ```js
> await deep.delete({
>   id: {
>      _gt: migrationsEndLink.id,
>   },
>   _not: {
>      type_id: 
>      up: {
>         tree_id: {
>            _eq: await deep.id("@deep-foundation/core", "containTree")
>         },
>         parent: {
>             type_id: await deep.id("@deep-foundation/core", "Package"),
>         }
>      },
>   }
> },
> ```
Yeah. This is something like what I need. empty type_id property  is redundant. Thank you

> Be sure to also keep PackageNamespace, PackageActive, PackageVersion links and so on.
I have forgot about that. Thank you

**Reply by konard** [reply]
I updated the code.
